rasa/shared/importers/importer.py

```diff
@@ -164,7 +164,9 @@ def _importer_from_dict(
         if module_path == RasaFileImporter.__name__:
             importer_class = RasaFileImporter
         elif module_path == MultiProjectImporter.__name__:
-            importer_class = MultiProjectImporter
+            # FIXME: we'd need to refactor the class hierarchy from
```

Comment:
why? Don't they use the same interface?

Reply:
`TrainingDataImporter` is both used as an interface and as some sort of a factory. I think this is the underlying issue here. I've tried to give a hint of `Type["TrainingDataImporter"]` to `importer_class` at line 165 (similar to what I did for brokers, etc...) and I get `Too many arguments for "TrainingDataImporter"  [call-arg]` when trying to instantiate the class. `TrainingDataImporter.__init__` doesn't exist

Reply:
Oh, I see what the issue is. We should have a `create` classmethod on the importers instead of using the constructor.

You could do

```python
 importer_class: Type["TrainingDataImporter"]
        if module_path == RasaFileImporter.__name__:
            importer_class = RasaFileImporter
        elif module_path == MultiProjectImporter.__name__:
            importer_class = MultiProjectImporter
        else:
            try:
                importer_class = cast(
                    Type["TrainingDataImporter"],
                    rasa.shared.utils.common.class_from_module_path(module_path),
                )
            except (AttributeError, ImportError):
                logging.warning(f"Importer '{module_path}' not found.")
                return None

        constructor_arguments = rasa.shared.utils.common.minimal_kwargs(
            importer_config, importer_class
        )

        return importer_class(
            config_path, domain_path, training_data_paths, **constructor_arguments
        )
```

and then add a constructor to `TrainingDataImporter` ala

```python
    def __init__(
        self,
        config_file: Optional[Text] = None,
        domain_path: Optional[Text] = None,
        training_data_paths: Optional[Union[List[Text], Text]] = None,
        **kwargs: Any
    ) -> None:
       raise NotImplementedError
```

Reply:
I tried and it doesn't break (pretty sure because of the `kwargs` part) 🤔

Reply:
indeed 🤷🏻 I don't know why it didn't work the first time (I might have stopped halfway through) 9f0aed66e91